File: packages/silo-client/src/transport/api-path.ts
/**
 * Every path in the client, built in one place. Every dynamic segment goes
 * through `encodeURIComponent`, so a name containing a slash or a space
 * addresses one resource rather than being split across the path.
 *
 * Uses `/envs/`, not `/environments/` — the guide says both are accepted
 *, and one spelling is enough to build with.
 */
export class ApiPath {
  static health(): string {
    return "/api/health";
  }

  static projects(): string {
    return "/api/projects";
  }

  static project(project: string): string {
    return `/api/projects/${encodeURIComponent(project)}`;
  }

  static environments(project: string): string {
    return `${ApiPath.project(project)}/envs`;
  }

  static environment(project: string, env: string): string {
    return `${ApiPath.environments(project)}/${encodeURIComponent(env)}`;
  }

  static projectVariables(project: string): string {
    return `${ApiPath.project(project)}/variables`;
  }

  static projectVariable(project: string, name: string): string {
    return `${ApiPath.projectVariables(project)}/${encodeURIComponent(name)}`;
  }

  static environmentVariables(project: string, env: string): string {
    return `${ApiPath.environment(project, env)}/variables`;
  }

  static environmentVariable(project: string, env: string, name: string): string {
    return `${ApiPath.environmentVariables(project, env)}/${encodeURIComponent(name)}`;
  }

  static collections(project: string, env: string): string {
    return `${ApiPath.environment(project, env)}/collections`;
  }

  static collection(project: string, env: string, name: string): string {
    return `${ApiPath.collections(project, env)}/${encodeURIComponent(name)}`;
  }

  static schemas(project: string, env: string): string {
    return `${ApiPath.environment(project, env)}/schemas`;
  }

  static collectionSchema(project: string, env: string, name: string): string {
    return `${ApiPath.collection(project, env, name)}/schema`;
  }

  /** List/create entries: the same address as `collection`, one level down
   * in what it means — the collection's own record versus its rows. */
  static entries(project: string, env: string, name: string): string {
    return ApiPath.collection(project, env, name);
  }

  static entry(project: string, env: string, name: string, id: string): string {
    return `${ApiPath.collection(project, env, name)}/${encodeURIComponent(id)}`;
  }

  static collectionSearch(project: string, env: string, name: string): string {
    return `${ApiPath.collection(project, env, name)}/search`;
  }

  static environmentSearch(project: string, env: string): string {
    return `${ApiPath.environment(project, env)}/search`;
  }

  static instanceSearch(): string {
    return "/api/search";
  }

  static media(): string {
    return "/api/media";
  }

  static mediaExtensions(): string {
    return "/api/media/extensions";
  }

  static mediaAsset(id: string): string {
    return `/api/media/${encodeURIComponent(id)}`;
  }

  static mediaAssetUsages(id: string): string {
    return `${ApiPath.mediaAsset(id)}/usages`;
  }

  static mediaBulkDelete(): string {
    return "/api/media/delete";
  }

  static mediaFolders(): string {
    return "/api/media/folders";
  }
}

File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/transport/ApiPath.java
package in.org.quicko.silo.client.transport;

import java.nio.charset.StandardCharsets;

/**
 * Every path in the client, built in one place. Every dynamic segment is
 * percent-encoded, so a name containing a slash or a space addresses one
 * resource rather than being split across the path.
 *
 * <p>Uses {@code /envs/}, not {@code /environments/} — the API accepts both, and
 * one spelling is enough to build with.
 */
public final class ApiPath {
  private ApiPath() {}

  public static String health() {
    return "/api/health";
  }

  public static String projects() {
    return "/api/projects";
  }

  public static String project(String project) {
    return "/api/projects/" + segment(project);
  }

  public static String environments(String project) {
    return project(project) + "/envs";
  }

  public static String environment(String project, String env) {
    return environments(project) + "/" + segment(env);
  }

  public static String projectVariables(String project) {
    return project(project) + "/variables";
  }

  public static String projectVariable(String project, String name) {
    return projectVariables(project) + "/" + segment(name);
  }

  public static String environmentVariables(String project, String env) {
    return environment(project, env) + "/variables";
  }

  public static String environmentVariable(String project, String env, String name) {
    return environmentVariables(project, env) + "/" + segment(name);
  }

  public static String collections(String project, String env) {
    return environment(project, env) + "/collections";
  }

  public static String collection(String project, String env, String name) {
    return collections(project, env) + "/" + segment(name);
  }

  public static String schemas(String project, String env) {
    return environment(project, env) + "/schemas";
  }

  public static String collectionSchema(String project, String env, String name) {
    return collection(project, env, name) + "/schema";
  }

  /**
   * List and create entries: the same address as {@link #collection}, one level
   * down in what it means — the collection's own record versus its rows.
   */
  public static String entries(String project, String env, String name) {
    return collection(project, env, name);
  }

  public static String entry(String project, String env, String name, String id) {
    return collection(project, env, name) + "/" + segment(id);
  }

  public static String collectionSearch(String project, String env, String name) {
    return collection(project, env, name) + "/search";
  }

  public static String environmentSearch(String project, String env) {
    return environment(project, env) + "/search";
  }

  public static String instanceSearch() {
    return "/api/search";
  }

  public static String media() {
    return "/api/media";
  }

  public static String mediaExtensions() {
    return "/api/media/extensions";
  }

  public static String mediaAsset(String id) {
    return "/api/media/" + segment(id);
  }

  public static String mediaAssetUsages(String id) {
    return mediaAsset(id) + "/usages";
  }

  public static String mediaAssetContent(String id) {
    return mediaAsset(id) + "/content";
  }

  public static String mediaBulkDelete() {
    return "/api/media/delete";
  }

  public static String mediaFolders() {
    return "/api/media/folders";
  }

  /**
   * Percent-encodes one path segment or query component. The unreserved set is
   * JavaScript's encodeURIComponent, so this client and the TypeScript one
   * address a name holding a space or a slash identically.
   */
  public static String segment(String value) {
    StringBuilder encoded = new StringBuilder(value.length() + 8);
    for (byte raw : value.getBytes(StandardCharsets.UTF_8)) {
      int octet = raw & 0xFF;
      if (isUnreserved(octet)) {
        encoded.append((char) octet);
      } else {
        encoded.append('%')
            .append(Character.toUpperCase(Character.forDigit(octet >> 4, 16)))
            .append(Character.toUpperCase(Character.forDigit(octet & 0x0F, 16)));
      }
    }
    return encoded.toString();
  }

  private static boolean isUnreserved(int octet) {
    return (octet >= 'A' && octet <= 'Z')
        || (octet >= 'a' && octet <= 'z')
        || (octet >= '0' && octet <= '9')
        || "-_.!~*'()".indexOf(octet) >= 0;
  }
}
